pilot/pkg/model/jwks_resolver.go

```diff
@@ -220,63 +222,40 @@ func newJwksResolverWithCABundlePaths(
-
-// GetPublicKey gets JWT public key and cache the key for future use.
-func (r *JwksResolver) GetPublicKey(issuer string, jwksURI string) (string, error) {
+// GetOrUpdatePublicKey returns the  JWT public key if it is available in the cache
+// or adds the jwksURI in the cache to fetch the public key in the background process
+func (r *JwksResolver) GetOrUpdatePublicKey(issuer string, jwksURI string) string {
 	now := time.Now()
 	key := jwtKey{issuer: issuer, jwksURI: jwksURI}
 	if val, found := r.keyEntries.Load(key); found {
 		e := val.(jwtPubKeyEntry)
 		// Update cached key's last used time.
 		e.lastUsedTime = now
 		r.keyEntries.Store(key, e)
-		if e.pubKey == "" {
-			return e.pubKey, errEmptyPubKeyFoundInCache
-		}
-		return e.pubKey, nil
+		return e.pubKey
 	}
-
-	var err error
-	var pubKey string
-	if jwksURI == "" {
-		// Fetch the jwks URI if it is not hardcoded on config.
-		jwksURI, err = r.resolveJwksURIUsingOpenID(issuer)
-	}
-	if err != nil {
-		log.Errorf("Failed to jwks URI from %q: %v", issuer, err)
-	} else {
-		var resp []byte
-		resp, err = r.getRemoteContentWithRetry(jwksURI, networkFetchRetryCountOnMainFlow)
-		if err != nil {
-			log.Errorf("Failed to fetch public key from %q: %v", jwksURI, err)
-		}
-		pubKey = string(resp)
-	}
-
 	r.keyEntries.Store(key, jwtPubKeyEntry{
-		pubKey:            pubKey,
+		pubKey:            "",
 		lastRefreshedTime: now,
 		lastUsedTime:      now,
 	})
-
-	return pubKey, err
+	// fetching the public key in the background
+	jwksuriChannel <- key
+	return ""
 }
 
 // BuildLocalJwks builds local Jwks by fetching the Jwt Public Key from the URL passed if it is empty.
 func (r *JwksResolver) BuildLocalJwks(jwksURI, jwtIssuer, jwtPubKey string) *envoy_jwt.JwtProvider_LocalJwks {
 	if jwtPubKey == "" {
-		var err error
-
 		// jwtKeyResolver should never be nil since the function is only called in Discovery Server request processing
 		// workflow, where the JWT key resolver should have already been initialized on server creation.
-		jwtPubKey, err = r.GetPublicKey(jwtIssuer, jwksURI)
-		if err != nil {
-			log.Errorf("Failed to fetch jwt public key from issuer %q, jwks uri %q: %s", jwtIssuer, jwksURI, err)
-			// This is a temporary workaround to reject a request with JWT token by using a fake jwks when istiod failed to fetch it.
-			// TODO(xulingqing): Find a better way to reject the request without using the fake jwks.
-			jwtPubKey = CreateFakeJwks(jwksURI)
-		}
+		jwtPubKey = r.GetOrUpdatePublicKey(jwtIssuer, jwksURI)
+	}
+	if jwtPubKey == "" {
+		log.Infof("The JWKS key is not yet fetched for issuer %s (%s), using a fake JWKS for now", jwtIssuer, jwksURI)
+		// This is a temporary workaround to reject a request with JWT token by using a fake jwks when istiod failed to fetch it.
+		// TODO(xulingqing): Find a better way to reject the request without using the fake jwks.
+		jwtPubKey = CreateFakeJwks(jwksURI)
```

Comment:
Are we sure that this is the right user facing behavior?

This means that if I apply some PeerAuthentication, my traffic will immediately break for a bit before it fetches later?